DAOS-16908 object: firewall support

jxiong opened the pull request:
Introduces support for DAOS clients operating behind a firewall. Previously, client-side firewalls had to be disabled as servers initiated connections. This feature provides a robust solution for clients operating in cloud environments.

This feature can be optionally enabled by setting `client_behind_firewall: true` in `daos_server.yaml` (disabled by default). When enabled, servers will no longer attempt to connect back to clients. Instead, they return a `DER_RECONNECT` error, prompting clients to initiate new connections to involved servers via ping RPC, and then retry the original RPC.

Leverages recent enhancements in underlying communication libraries:
- **libfabric:** PRs # 10912, # 10922, # 10949, # 10955
- **Mercury:** PRs # 787, # 793

## Test performed:
### compatibility and function tests
- the test involves 3 server instances with:
  - "Old Server": instances without this PR
  - "New Server w/  firewall": instance with this PR and `client_hehind_client` set to `true` in daos_server.yml
  - "New Server w/o firewall": instance with this PR and `client_hehind_client` set to `false` in daos_server.yml
- meanwhile, we started 2 clients: 
  - "Old Client": client doesn't have this PR
  - "New Client": client with this PR
- test is performed by simple fio via dfuse mountpoint
- the test results are as follows:
```
+-------------------------+------------+------------+
|                         | Old Client | New Client | 
+-------------------------+------------+------------+
| Old Server              | passed     | passed     | 
| New Server w/  firewall | passed     | passed     | 
| New Server w/o firewall | passed     | passed     | 
+-------------------------+------------+------------+
```
In the tests, we made sure that all fio runs can be finished successfully. And in the combination of "New Client" and "New Server w/  firewall", servers won't initiate any connecting request to the client. In particular, the iptables rule `iptables -A INPUT -s <server_ip> -p tcp --syn -j DROP` is applied on the client to block any TCP syn packet from servers.
- we also did the compatibility test with daos_server.yml to make sure the new software can recognize old version of daos_server.yml and `client_hehind_firewall` is indeed set to `false` by default. However, old version of software will run into errors if the daos_server.yml has `client_behind_firewall` setting because it didn't understand the option.

Signed-off-by: Jinshan Xiong <jinshanx@google.com>
Signed-off-by: Yokesh Jayakumar <karthj@google.com>
Signed-off-by: Jeff Olivier <jeffolivier@google.com>

### Steps for the author:

* [x] Commit message follows the [guidelines](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments).
* [x] Appropriate [Features or Test-tag](https://daosio.atlassian.net/wiki/spaces/DC/pages/10984259629/Test+Tags) pragmas were used.
* [x] Appropriate [Functional Test Stages](https://daosio.atlassian.net/wiki/spaces/DC/pages/12147556353/CI+Functional+Test+Stages) were run.
* [x] At least two positive code reviews including at least one code owner from each category referenced in the PR.
* [x] Testing is complete. If necessary, forced-landing label added and a reason added in a comment.

#### After all prior steps are complete:
* [x] Gatekeeper requested (daos-gatekeeper added as a reviewer).

jolivier23 committed:
Merge pull request #16120 from daos-stack/karthj/feature/firewall

Merge karthj/feature/firewall into feature/firewall

karthjyojay committed:
DAOS-16908 object: add client-side target pinging on update retry (#16086)

In the case of update bulk transfers, the server targets may try to communicate with clients who have not yet established a connection with them. In this case, the connection operation hangs forever. In Google Parallelstore, this is an issue because we do not have control over the client firewalls. To deal with this issue, we will have the server tell the client that they cannot connect to the client due to this scenario and the client will ping the affected targets before retrying their update RPC.

The purpose of this PR is to introduce client-side logic to ping targets involved in an update RPC.

Signed-off-by: Yokesh Jayakumar <karthj@google.com>

karthjyojay committed:
Merge master into karthj/feature/firewall

Signed-off-by: Yokesh Jayakumar <karthj@google.com>

jolivier23 committed:
Merge pull request #16211 from daos-stack/karthj/feature/firewall

Merge latest master into feature/firewall

karthjyojay committed:
Merge branch 'master' into feature/firewall

karthjyojay committed:
DAOS-16908 object: add client-side compound RPC pinging RPC on DER_RE… (#16246)

* DAOS-16908 object: add client-side compound RPC pinging RPC on DER_RECONNECT.

Like with the single update case, when the server returns DER_RECONNECT, it means
that the server did not have an established TCP connection to the client it tried
to do the bulk transfer from. We need to handle this logic in the compound RPC
case, where each subrequest could result in this error.

Signed-off-by: Yokesh Jayakumar <karthj@google.com>

* Resolve commit feedback

Signed-off-by: Yokesh Jayakumar <karthj@google.com>

* Fix naming of test

Signed-off-by: Yokesh Jayakumar <karthj@google.com>

* Move ordering of DAOS_FAIL_CHECK because we want the actual bulk transfer for the update to trigger DAOS_FAIL_CHECK

Signed-off-by: Yokesh Jayakumar <karthj@google.com>

* Refactor layers of abstraction between task and individual target ping tasks

Signed-off-by: Yokesh Jayakumar <karthj@google.com>

* Fix linting error

Signed-off-by: Yokesh Jayakumar <karthj@google.com>

* Initialize rc to 0 and wrap TX_FAILED in else

Signed-off-by: Yokesh Jayakumar <karthj@google.com>

* Trigger pr tests

Signed-off-by: Yokesh Jayakumar <karthj@google.com>

* Resolve commit feedback by making unit tests more memory efficient and add
a test case for when the compound body bulk transfer may throw a DER_RECONNECT.

Signed-off-by: Yokesh Jayakumar <karthj@google.com>

* Add par_barrier to new unit tests

Signed-off-by: Yokesh Jayakumar <karthj@google.com>

* Make compound RPC retry on DER_RECONNECT instead of failing transaction

Signed-off-by: Yokesh Jayakumar <karthj@google.com>

* Rename enum for target dtx enum and merge crt_bulk_get if statement

Signed-off-by: Yokesh Jayakumar <karthj@google.com>

* Refactor common code to return DER_RECONNECT

Signed-off-by: Yokesh Jayakumar <karthj@google.com>

* Resolve more feedback

Signed-off-by: Yokesh Jayakumar <karthj@google.com>

* Resolve clang-format

Signed-off-by: Yokesh Jayakumar <karthj@google.com>

* Determine whether to fault injection in caller of obj_bulk_transfer

Signed-off-by: Yokesh Jayakumar <karthj@google.com>

---------

Signed-off-by: Yokesh Jayakumar <karthj@google.com>

jolivier23 committed:
Merge remote-tracking branch 'comm/master' into feature/firewall

karthjyojay committed:
DAOS-16908 control: add flag for TCP firewall feature in control plane (#16292)

* Add flag for TCP firewall feature in control plane.

This flag will be set in the daos_server.yml and sent to the client via the
daos_agent. If the flag is true, we will set a client-side environment variable
that will be used by Mercury and Libfabric to set a bit in subsequent RPC requests
to the server. If the flag is false, it will not be set.

Signed-off-by: Yokesh Jayakumar <karthj@google.com>

* Trigger tests

Signed-off-by: Yokesh Jayakumar <karthj@google.com>

* Rename flag, make it default to true, and resolve other feedback

Signed-off-by: Yokesh Jayakumar <karthj@google.com>

* Make DisableClientFirewallMode a top-down server config

Signed-off-by: Yokesh Jayakumar <karthj@google.com>

* Free client_firewall_mode string

Signed-off-by: Yokesh Jayakumar <karthj@google.com>

* Resolve commit feedback.

Signed-off-by: Yokesh Jayakumar <karthj@google.com>

* Remove debugging print statement in server_test

Signed-off-by: Yokesh Jayakumar <karthj@google.com>

* Update documentation in daos_server.yml

Signed-off-by: Yokesh Jayakumar <karthj@google.com>

---------

Signed-off-by: Yokesh Jayakumar <karthj@google.com>

jolivier23 committed:
DAOS-623 common: Add thread_private fail_loc (#16248)

Add threadprivate flag capability

Signed-off-by: Jeff Olivier <jeffolivier@google.com>

jxiong committed:
DAOS-16908 cart: Propagate firewall error to daos (#16428)

This PR parses HG_HOSTFIREWALL error from mercury to DER_RECONNECT so
that the client will know that the RPC is blocked by firewall.

Signed-off-by: Jinshan Xiong <jinshanx@google.com>

jxiong committed:
Merge remote-tracking branch 'origin/feature/firewall'

daosbuild3 commented:
Test stage Functional Hardware Medium Verbs Provider MD on SSD completed with status FAILURE.  https://jenkins-3.daos.hpc.amslabs.hpecorp.net//job/daos-stack/job/daos/view/change-requests/job/PR-16429/2/execution/node/1427/log

jolivier23 commented:
> Introduces support for DAOS clients operating behind a firewall. Previously, client-side firewalls had to be disabled as servers initiated connections. This feature provides a robust solution for clients operating in cloud environments.
> 
> This feature can be optionally enabled by setting `disable_client_firewall_mode: true` in `daos_server.yaml` (disabled by default). When enabled, servers will no longer attempt to connect back to clients. Instead, they return a `DER_RECONNECT` error, prompting clients to initiate new connections to involved servers via ping RPC, and then retry the original RPC.
> 
> Leverages recent enhancements in underlying communication libraries:
> 
> * **libfabric:** PRs # 10912, # 10922, # 10949, # 10955
> * **Mercury:** PRs # 787, # 793
> 
> Signed-off-by: Jinshan Xiong [jinshanx@google.com](mailto:jinshanx@google.com) Signed-off-by: Yokesh Jayakumar [karthj@google.com](mailto:karthj@google.com) Signed-off-by: Jeff Olivier [jeffolivier@google.com](mailto:jeffolivier@google.com)
> 
> ### Steps for the author:
> * [ ]  Commit message follows the [guidelines](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments).
> * [ ]  Appropriate [Features or Test-tag](https://daosio.atlassian.net/wiki/spaces/DC/pages/10984259629/Test+Tags) pragmas were used.
> * [ ]  Appropriate [Functional Test Stages](https://daosio.atlassian.net/wiki/spaces/DC/pages/12147556353/CI+Functional+Test+Stages) were run.
> * [ ]  At least two positive code reviews including at least one code owner from each category referenced in the PR.
> * [ ]  Testing is complete. If necessary, forced-landing label added and a reason added in a comment.
> 
> #### After all prior steps are complete:
> * [ ]  Gatekeeper requested (daos-gatekeeper added as a reviewer).

I think we should change the flag name.  I find disable_client_firewall_mode to be incredibly difficult for my brain to wrap around.

What about 

`assume_firewall: true`

jxiong committed:
DAOS-16908 control: change the firewall config name (#16448)

Change the firewall mode name from `disable_client_firewall_mode`
to `client_behind_firewall` in daos_server.yml.

Signed-off-by: Jinshan Xiong <jinshanx@google.com>

jxiong committed:
Merge remote-tracking branch 'origin/master' into feature/firewall

jxiong committed:
fix linting

Signed-off-by: Jinshan Xiong <jinshanx@google.com>

soumagne commented on src/cart/crt_hg.h:
```diff
@@ -189,36 +189,8 @@ void crt_provider_get_ctx_list_and_num(bool primary, crt_provider_t provider, d_
-	};
-}
+int
+crt_hgret_2_der(int hg_ret);
```
what is the reason for making that one non inline and leave the next one inline ?

jxiong replied:
This is because we want daos to compile even with old version of mercury where `HG_HOSTFIREWALL` is not defined. Therefore, we enhanced scons to detect it and define one if it's not defined: https://github.com/daos-stack/daos/pull/16429/files#diff-433ec396346afa932f0d8f67525073e133a0e3dbb92b15f9503e81574a2e3e36R113.

The reason to make it non-inline is to move it out of header file. In that case, we can limit the definition above limited to files under cart directory. Otherwise, `-DHG_HOSTFIREWALL=0xdeadbeaf` has to be applied all files where `crg_hg.h` is used.

tanabarr commented on src/control/lib/control/network.go:
```diff
@@ -216,14 +217,15 @@ type (
+		SrvSrxSet          int32    `json:"srv_srx_set"`
+		EnvVars            []string `json:"env_vars"`
+		ProviderIdx        uint32   `json:"provider_idx"`
+		ClientFirewallMode bool     `json:"client_firewall_mode"`
```
I wouldn't block on it but I think it's easier for people to follow if the field name matches the flag used in the command. This should generally be followed unless there is a specific reason not to IMO.

tanabarr commented on src/control/server/config/server.go:
```diff
@@ -48,26 +49,27 @@ type deprecatedParams struct {
+	CoreDumpFilter     uint8                     `yaml:"core_dump_filter,omitempty"`
+	ClientEnvVars      []string                  `yaml:"client_env_vars,omitempty"`
+	SupportConfig      SupportConfig             `yaml:"support_config,omitempty"`
+	ClientFirewallMode *bool                     `yaml:"client_behind_firewall,omitempty"`
```
This kind of inconsistency will cause problems for people trying to follow the code. The YAML name should match the field name unless there is a specific reason not to and IMO that should follow the flag used in the command.

jxiong replied:
`client_behind_firewall` is user facing, `ClientFirewallMode` is programmer facing. Anyways, it will lead to large changes if we need to modify it, and we have landed the PR internally already, so let's keep it as-is

kjacque replied:
I can live with it at other levels where we use the "client firewall mode" terminology, but I do think at the config level it is good to have the parameter match the name.

tanabarr commented on src/control/server/config/server.go:
```diff
@@ -326,6 +334,8 @@ func (cfg *Server) WithTelemetryPort(port int) *Server {
 // DefaultServer creates a new instance of configuration struct
 // populated with defaults.
 func DefaultServer() *Server {
+	clientFirewallMode := false
```
There isn't anything necessarily wrong with doing it this way but in other areas where we are trying to do something similar this is set during validation e.g. VMD related settings.

tanabarr commented on src/control/server/config/server_test.go:
```diff
@@ -245,6 +246,7 @@ func TestServerConfig_Constructed(t *testing.T) {
 		WithDisableVFIO(true).   // vfio enabled by default
 		WithDisableVMD(true).    // vmd enabled by default
 		WithEnableHotplug(true). // hotplug disabled by default
+		WithClientFirewallMode(true).
```
nit: could add similar comment as above for clarity

jxiong replied:
I think we should delete all the comments above instead. The comments didn't add any useful information, lol

tanabarr reviewed (dismissed):
LGTM, only nonblocking comments that could be addressed at the author's discretion.

jxiong committed:
Merge branch 'master' into jxiong/daos_16908

Signed-off-by: Jinshan Xiong <jinshan.xiong@gmail.com>

kjacque reviewed (changes requested):
Some minor style comments that should be addressed before landing. I agree with Tom that we need consistency in the names on the control plane side. Otherwise looks OK from the control plane perspective.

I gave the C side a more cursory review and leave the details to be reviewed by someone more knowledgable in the object and client code.

jxiong commented on src/include/daos_srv/pool.h:
```diff
@@ -30,6 +31,10 @@
 struct ds_pool_svc;
 
+/* This is used to ensure in unit testing that clients ping targets after receiving a DER_RECONNECT
+error from the server in the case there is fault injection to return such an error.*/
```
Done

jxiong commented on src/mgmt/cli_mgmt.c:
```diff
@@ -653,6 +655,18 @@ dc_mgmt_net_cfg_init(const char *name, crt_init_options_t *crt_info)
 	}
 
+	if (info->client_firewall_mode) {
+		rc = asprintf(&client_firewall_mode, "%d", info->client_firewall_mode);
```
Done

jxiong commented on src/object/cli_obj.c:
```diff
@@ -1404,6 +1405,123 @@ obj_pool_query_task(tse_sched_t *sched, struct dc_object *obj,
+	int                       rc;
+	struct ping_tgt_task_arg *arg = tse_task_buf_embedded(task, sizeof(*arg));
+	pool_hdl                      = arg->pool_hdl;
+	tgt_id                        = arg->tgt_id;
```
Done

jxiong committed:
fix per comments

daosbuild3 commented:
Test stage Unit Test on EL 8.8 completed with status UNSTABLE.  https://jenkins-3.daos.hpc.amslabs.hpecorp.net/job/daos-stack/job/daos//view/change-requests/job/PR-16429/6/testReport/

tanabarr reviewed (changes requested):
One small comment fix required, otherwise LGTM. I'm not sure why we reverted from the pointer but I don't feel strongly enough about it to object.

kjacque commented on src/control/server/config/server_test.go:
```diff
@@ -244,10 +244,10 @@ func TestServerConfig_Constructed(t *testing.T) {
+		WithDisableVFIO(true).        // vfio enabled by default
+		WithDisableVMD(true).         // vmd enabled by default
+		WithEnableHotplug(true).      // hotplug disabled by default
+		WithClientFirewallMode(true). // firewall enabled by default
```
+1, this is incorrect now

jxiong committed:
change the comment

jolivier23 commented on src/include/daos_srv/pool.h:
```diff
@@ -30,6 +31,12 @@
+ * receiving a DER_RECONNECT error from the server in the case there is fault
+ * injection to return such an error.
+ */
+extern __thread bool firewall_blocking_flag;
```
This can be removed.  It's no longer used because threadprivate support was added to FAIL_LOC

jolivier23 commented on src/pool/srv_target.c:
```diff
@@ -41,6 +42,8 @@
 #include "rpc.h"
 #include "srv_internal.h"
 
+__thread bool firewall_blocking_flag = false;
```
Same comment, can be removed.

jxiong replied:
Done.

jxiong committed:
remove unused firewall_blocking_flag

daosbuild3 commented:
Test stage Unit Test on EL 8.8 completed with status UNSTABLE.  https://jenkins-3.daos.hpc.amslabs.hpecorp.net/job/daos-stack/job/daos//view/change-requests/job/PR-16429/9/testReport/

jxiong committed:
Merge branch 'master' into jxiong/daos_16908

Signed-off-by: Jinshan Xiong <jinshan.xiong@gmail.com>

jxiong committed:
Merge branch 'master' into jxiong/daos_16908

kjacque reviewed (dismissed):
Still don't love the different internal/external names for this feature, but I'll live with it. (are we sure "client_firewall" wouldn't suffice as a happy medium?)

jxiong commented on src/control/server/config/server.go:
```diff
@@ -78,6 +79,7 @@ type Server struct {
 	CoreDumpFilter     uint8                     `yaml:"core_dump_filter,omitempty"`
 	ClientEnvVars      []string                  `yaml:"client_env_vars,omitempty"`
 	SupportConfig      SupportConfig             `yaml:"support_config,omitempty"`
+	ClientFirewallMode bool                      `yaml:"client_behind_firewall,omitempty"`
```
nope. That'd lead to a big change.

jxiong commented:
> Still don't love the different internal/external names for this feature, but I'll live with it. (are we sure "client_firewall" wouldn't suffice as a happy medium?)

Thanks for the suggestion. I always prefer short names if possible. However, `client_firewall` does not seem to be a right term. `client_behind_firewall` is carefully chosen and I believe there is no ambiguity when a user sees this word in the configuration file. Originally it came with a *very* confusing name and that's why there exists a different name in user-facing term and implementation.

daltonbohning commented:
This is failing build stages, possibly due to recent CI issues, so I restarted Jenkins testing

kjacque commented:
@jxiong Looks like CI failed again. Can you merge the latest master and repush please?

jxiong committed:
Merge remote-tracking branch 'origin/master' into jxiong/daos_16908

daosbuild3 commented:
Test stage NLT on EL 8.8 completed with status UNSTABLE.  https://jenkins-3.daos.hpc.amslabs.hpecorp.net/job/daos-stack/job/daos//view/change-requests/job/PR-16429/13/testReport/

daltonbohning commented:
This has not finished CI testing so cannot be merged yet

jxiong committed:
client_behind_firewall is omitted if not set

jxiong committed:
change WithEnableHotplug to WithDisableHotplug

daosbuild3 commented:
Test stage NLT on EL 8.8 completed with status UNSTABLE.  https://jenkins-3.daos.hpc.amslabs.hpecorp.net/job/daos-stack/job/daos//view/change-requests/job/PR-16429/15/testReport/

jxiong committed:
Merge remote-tracking branch 'origin/master' into jxiong/daos_16908

daosbuild3 commented:
Test stage Unit Test bdev with memcheck on EL 8.8 completed with status FAILURE.  https://jenkins-3.daos.hpc.amslabs.hpecorp.net/job/daos-stack/job/daos/job/PR-16429/16/display/redirect

daosbuild3 commented:
Test stage Unit Test on EL 8.8 completed with status FAILURE.  https://jenkins-3.daos.hpc.amslabs.hpecorp.net/job/daos-stack/job/daos/job/PR-16429/16/display/redirect

jolivier23 committed:
Merge remote-tracking branch 'comm/master' into jxiong/daos_16908

Signed-off-by: Jeff Olivier <jeffolivier@google.com>

jolivier23 committed:
Allow-unstable-test: true

Signed-off-by: Jeff Olivier <jeffolivier@google.com>

jolivier23 committed:
Remove protoc changes

Allow-unstable-test: true

Signed-off-by: Jeff Olivier <jeffolivier@google.com>

jolivier23 committed:
Allow-unstable-test: true

Change-Id: Ifdd97c33e9c9df95600a3758b14cbc2656d952de
Signed-off-by: Jeff Olivier <jeffolivier@google.com>

daosbuild3 commented:
Test stage Build on EL 9 completed with status FAILURE.  https://jenkins-3.daos.hpc.amslabs.hpecorp.net//job/daos-stack/job/daos/view/change-requests/job/PR-16429/20/execution/node/360/log

daosbuild3 commented:
Test stage Build RPM on EL 9 completed with status FAILURE.  https://jenkins-3.daos.hpc.amslabs.hpecorp.net//job/daos-stack/job/daos/view/change-requests/job/PR-16429/20/execution/node/345/log

daosbuild3 commented:
Test stage Build RPM on EL 8 completed with status FAILURE.  https://jenkins-3.daos.hpc.amslabs.hpecorp.net//job/daos-stack/job/daos/view/change-requests/job/PR-16429/20/execution/node/351/log

jolivier23 committed:
Merge remote-tracking branch 'comm/master' into jxiong/daos_16908

Change-Id: I1cc8bf2abe8474480aeb7759deed434bac4a03ed

jolivier23 committed:
Allow-unstable-test: true

Change-Id: I4105969a56c613278383fb8196291b9816e3ba7b
Signed-off-by: Jeff Olivier <jeffolivier@google.com>

daosbuild3 commented:
Test stage Build on EL 8.8 completed with status FAILURE.  https://jenkins-3.daos.hpc.amslabs.hpecorp.net//job/daos-stack/job/daos/view/change-requests/job/PR-16429/21/execution/node/353/log

jolivier23 committed:
Merge remote-tracking branch 'comm/master' into jxiong/daos_16908

Change-Id: Id1ec5e73fd936e12aaca2b9d1b599fb7b4a40205

jolivier23 committed:
Rebase with master

Allow-unstable-test: true
Change-Id: I599676685d4adfbe3f83ebe855c0d9c6bb0b153d
Signed-off-by: Jeff Olivier <jeffolivier@google.com>

jolivier23 committed:
Try again

Allow-unstable-test: true

Change-Id: Iea519b4faa6a3e66cd8f0759e308909fa015fe4a
Signed-off-by: Jeff Olivier <jeffolivier@google.com>

Commit:
Merge branch 'master' into jxiong/daos_16908

Signed-off-by: Michael MacDonald <mjmac@google.com>

jolivier23 committed:
Merge remote-tracking branch 'comm/master' into jxiong/daos_16908

Change-Id: Ia81dc02d70908cdda19e09d8fa6dfabfaf9f51c0

jolivier23 committed:
Empty commit after merge

Change-Id: If21196231389583629d6c76d38b9748668577fd4
Signed-off-by: Jeff Olivier <jeffolivier@google.com>

jxiong committed:
Merge remote-tracking branch 'master' into jxiong/daos_16908

jxiong committed:
linting fix

jxiong committed:
Merge branch 'master' into jxiong/daos_16908

Signed-off-by: Jinshan Xiong <jinshan.xiong@gmail.com>

daosbuild3 commented:
Test stage Build on Leap 15.5 with Intel-C and TARGET_PREFIX completed with status FAILURE.  https://jenkins-3.daos.hpc.amslabs.hpecorp.net//job/daos-stack/job/daos/view/change-requests/job/PR-16429/28/execution/node/294/log

daosbuild3 commented:
Test stage Build on Leap 15.5 completed with status FAILURE.  https://jenkins-3.daos.hpc.amslabs.hpecorp.net//job/daos-stack/job/daos/view/change-requests/job/PR-16429/28/execution/node/405/log

jxiong committed:
Refactor obj_retry_cb to include is_pool_task variable

Signed-off-by: Jinshan Xiong <jinshan.xiong@gmail.com>
